### ISSUE TITLE
Support showing "source"

### ISSUE THREAD

**Issue by mattgodbolt** [mattgodbolt/jsbeeb#107]
In conversation with Kieran, came up that an annotated source (e.g. MOS disassemblies, Exile disassemblies) would be fantastic to attach somehow to the debugger.

**Comment by richtw1** [comment]
I've thought for a while that a great feature for an emulator would be an intelligent (of sorts) disassembler built-in, much in the vein of the thing you once wrote in C#, Matt.  Key features would be:

* assigning metadata to bytes in memory (unknown, opcode, [self-modified] operand, data, etc). Obviously we can do this as we are actually executing code.
* attach symbol names (labels) to addresses, which are used where possible in the disassembly.  These would be initially populated from JSR/JMP/branches as much as possible with autogenerated names.
* coerce use of expressions containing symbol names for operands, e.g. `LDA #(update-1) AND 255`
* highlight blocks of code which have just been run (I could imagine this instrumental in identifying routines, e.g. reset highlight / do something / break and examine code)
* embed comments (either per instruction, or to insert before a given address)
* with reasonable enough heuristics, you could also pull out sprite data this way.

This would probably slow down emulation sufficiently that it might be worth generating a new block of emulation functions for use when 'debug mode' is turned on.

**Comment by mattgodbolt** [comment]
Right: the "highlight blocks of code just run" is actually pretty easy (within the last 256 instructions at least); I had a prototype but couldn't get the UI "nice".

What aspects would slow down the emulation (other than a more fine-grained "run recently" thing?)

All for this though!

**Comment by richtw1** [comment]
I can imagine just the incremental complexity of tracking more stuff would start to become significant (particularly more complex stuff like tracking where bytes come from which ultimately end up in screen memory).  Maybe it's nothing, I haven't seen performance on lower-end machines or mobile devices.  I guess I'm invoking the mantra of "if you don't use it, you shouldn't pay for it" here :)

**Comment by mattgodbolt** [comment]
Right; yeah -- I didn't realise you meant that level of tracking too but it does make sense! Tracking a per-byte "provenance" would be somewhat easy to do and could be hidden behind a never-taken "if" just like the tracepoints are.

That said; there's two halves to this feature request: taking existing disassemblies e.g. http://www.level7.org.uk/miscellany/exile-disassembly.txt and parsing and showing them, and then having the tools to be able to reverse engineer more easily within jsbeeb.

I know Kieran's doing a Thrust disassembly at the moment for example; and he's targeting a re-assemblable BeebAsm file. Any thoughts on what a good file format would be for all this?

**Comment by richtw1** [comment]
Also consider that paging messes things up a bit; the addresses used would have to be the flattened ones so they could all coexist.  There are other disassemblies too, e.g. BASIC, DFS, etc.

Re file format, no not sure.  I'd definitely lean towards something human readable, even some kind of strict standardized disassembly format might work.  Not XML though :laughing: 

The thing about being able to describe operands by expressions containing existing labels is an important one, as without that you can lose some of the connection between the source and the object code.  Again, heuristics (but ever more complicated ones) can help do that automatically.  Code like this is an obvious use:

    LDA #data MOD 256
    STA zp
    LDA #data DIV 256
    STA zp+1

But not so obvious is this:

    LDA #(update-1) DIV 256
    PHA
    LDA #(update-1) MOD 256  ; might be the other way round
    PHA
    RTS

Exile has an even more complicated encoding of the routine addresses which mixes other stuff in the bytes - ideally it'd be possible to mark up those tables with the precise expressions that generated them originally.

**Comment by mattgodbolt** [comment]
Right: re format. I was just re-reading the Exile disassembly, and its format is pretty machine-parseable as-is. So that could work.

And yeah; DIV/MOD stuff would require a lot of heuristic magic. But with enough "recognizers" we can probably spot them :)

The RTS case there could be picked up; the "provenance" of the address as tracked would be "immediate value from opcode LDA #xxx" and so on. Those themselves would be tracked of course from "loaded from disc" and or "loaded from disc; copied to this address" in the case of relocated code.

I reckon we're on to something. Now to find some spare time...

**Comment by richtw1** [comment]
I think it's certainly feasible (I thought about it lots when I was starting out on a C++ implementation of jsbeeb - never have been that happy with JS! - the idea being to include something exactly like this, but which would run on much less powerful machines, e.g. Raspberry Pi).

Re loading from disc, relocating etc.  I always just imagined a big RESET button to be honest, rather than trying to track the history of memory contents.  Reasonable workflow would be:

* load Exile disc image
* put breakpoint on 78ED
* boot disc
* when it breaks, RESET ALL THE THINGS
* ???
* profit!

**Comment by richtw1** [comment]
Not that there's any reason why all this stuff _couldn't_ be tracked of course. But then you have the multiple universes of distinct disassemblies, and I figured that the model of new/load/save works better in this case.

The tricky thing to detect heuristically would be things like adding the base address to a zp indirect address, stuff like:

    LDA num
    STA zp
    LDA #0
    ASL zp
    ROL A
    ASL zp
    ROL A
    ASL zp
    ROL A
    ADC #base
    STA zp+1

even worse if there's just a MSB added, like that.

**Comment by mattgodbolt** [comment]
Re: C++ jsbeeb; have you seen they've got a new beebem repo?

**Comment by mattgodbolt** [comment]
I went and looked at what already exists, so we consume something rather than invent something.
Summary below; happy to expand any of it.

**There is a de-facto standard, but only for the trivial layer.** BeebAsm's `-d`/`-dd` output (the
"Swift" format) is read by b2, b-em and BeebEm, i.e. every BBC emulator that has symbol support at
all. VICE `al` format is the second-place convergence (b2 and BeebEm). Those are the only two
formats more than one BBC emulator reads. For anything richer than name/address there is nothing at
all, and that is the part we actually want.

**What BeebAsm emits today**, verified by building it and running it rather than from the docs.
`-d`/`-dd` plus `-labels <file>` gives one line:

```
[{'start':8192L,'loop':8197L,'data':8200L,'end':8203L}]
```

Decimal addresses, `L` suffix, unquoted names. `-d` is global labels only with the leading `.`
stripped; `-dd` is all labels with the `.` kept and nested scopes as dotted paths (`.loop.inner`).
Two things worth knowing: `name = value` constants are never exported (`SymbolTable::Dump` filters
on `IsLabel()`), so `oswrch` and friends do not appear; and the `-v` listing drops all comments,
has no file/line info, and resolves symbolic operands away (`JSR osasci` lists as `JSR &FFE3`). So
BeebAsm cannot currently emit anything that would let us show annotated source. Note that #671's
suggestion to add a `--debug-info` flag is partly reinventing `-labels`, which has existed for
years and is used in real build scripts.

**What b2 does**, and more usefully why. Tom's notes in `doc/b2_notes.org` are explicit:

> Goal here is just to make the disassembly and trace output slightly easier to read [...] Non-goals:
> dealing with source files; anything that's at all complicated to do!; anything that's a UI
> rabbithole

So b2 has deliberately declined the annotated-source problem. It reads six formats (VICE, ACME,
BeebAsm, and three 64tass variants), each producing only `{uint16 address, name}`. No comment field.
The genuinely good idea to borrow is its paging model: a symbol *file* carries "address suffixes"
naming paged regions (`m` main, `o` OS, `s` shadow, `n` ANDY, `h` HAZEL, `0`-`f` sideways banks,
`p`/`r` parasite), in Exclusive or Inclusive mode, plus 256 numbered groups for bulk
enable/disable. That is a better answer to Rich's 2016 "the addresses would have to be flattened"
point than flattening, because it degrades gracefully. The symbol work was contributed by
@markjfisher in tom-seddon/b2#475, against tom-seddon/b2#129 which is now closed.

**The corpora are the constraint, and neither fits a flat map.**

Mark Moxon's repos are BeebAsm and really do assemble byte-identical (there is a `crc32.py` verify
step against reference binaries extracted from the real discs; all 27 files match). But the builds
pass `-v`, not `-labels`, so there is no symbol file. What there *is* is a committed `compile.txt`
in every repo (1.6 MB for Elite disc) holding the address/label listing, which the README
explicitly offers as a debugging reference. The killer detail: `elite-source-docked.asm` and
`elite-source-flight.asm` both `ORG &11E3` and share 823 label names at different addresses, and
`elite-ships-a.asm` through `-p.asm` are 16 overlays all at `&5600`. Across a single build, 758 of
1,766 label names (43%) resolve to more than one address. Also, his repos carry no licence and the
README says "No other use is permitted", so we should ask before shipping anything derived.

The Level 7 site is much bigger than I realised: 71 disassemblies, not just Exile, still being
updated (most recent July 2026). The format is a fixed-column plain-text listing, address at col 1,
mnemonic at col 16, prose comment at col 69, labels on their own line as `; name`. It parses
cleanly (I had 0.23% unparsed, and all of those are typos in the source). Two gotchas: `#` is both
the immediate prefix and the comment marker, and `;` is both the label marker and the
operand-annotation marker, so anchor on columns rather than splitting on whitespace. One caveat
that matters: the format changed in March 2025 (the old one had trailing colons on addresses and no
right-hand comments), which has already silently broken at least one downstream parser.

**The thing nobody in either issue mentioned**: @simondotm's beeb-vsc already embeds jsbeeb, has a
TypeScript reimplementation of BeebAsm, and does source-level debugging with breakpoints and named
watches. It emits a `<target>.map` JSON:

```ts
export type SourceMapFile = {
  sources:   { [id: number]: string }
  labels:    { [label: string]: number }
  symbols:   { [label: string]: number }
  addresses: { [address: number]: { file, line, column, parent } }
}
```

That is the right shape and it is the one shape neither #107 nor #671 proposed: store address to
(file, line) and point at the real source, rather than copying comment text into a sidecar. It
gives us Moxon's structured headers and Level 7's column-69 prose verbatim, with no lossy
extraction and no question about redistributing anyone's words.

**Proposal, in two independent layers.** I think the mistake in #671 is treating these as one
format.

1. *Symbols.* Adopt BeebAsm `-d`/`-dd` and VICE `al`. Roughly 20 lines of JS each, and it makes us
   interoperable with b2, b-em and BeebEm on day one. Nothing to invent here.
2. *Annotated source.* Our own format, modelled on beeb-vsc's source map (address to source and
   line, plus an overlay/group id), not on #671's `{addr, name, comment}` sidecar, which cannot
   represent either corpus.

**On discovery**: a curated index keyed on the disc, with manual attach still available, mapping to
pre-parsed data fetched on demand from a separate repo. Some notes on the specifics:

- We already hash discs. `discFor()` in `src/fdc.js` calls `setOriginalImageCrc32()` on every load
  path (local file, URL, STH, Google Drive), with the unzipped inner image bytes, at load time and
  before any write, and `src/main.js` already uses it for snapshot verification. So the "what gets
  hashed, and is it stable" question is answered by shipped code. It is CRC32 rather than a
  cryptographic hash, so the index should treat it as a hint and the payload should carry enough
  identifying info to sanity-check.
- The index should live in a separate curated repo and be fetched, not bundled. Elite's
  `compile.txt` alone is 1.6 MB and the Level 7 Exile text is 875 KB; bundling one game would rival
  our whole payload. The CRC-to-title-and-URL index is small enough to fetch eagerly, the
  annotations must be lazy.
- It should map to pre-parsed data produced at build time by a converter in that repo, not to raw
  upstream files parsed in the browser. Level 7 changing format in March 2025 is the argument: a
  build-time converter fails loudly in CI, browser parsing fails silently for a user. It also keeps
  us pointing at Moxon's repos rather than redistributing them.
- Manual attach (`?annotations=<url>`, plus drag-and-drop) goes through the same parser, so it does
  not rot. That is also the path a developer uses to point at the `.labels` their own build just
  produced.
- Because we control the index and can pre-parse, the on-the-wire format is ours and the question
  stops being "which assembler format do we adopt" and becomes "which formats does the converter
  ingest", which we can extend over time at no runtime cost.

**Suggested first target: Level 7's Exile, not Elite.** One file, parses cleanly, the author states
"no additional copyright restrictions beyond those already present on the game itself", there is an
existing byte-verified BeebAsm rebuild to check against (tom-seddon/exile_disassembly), and it is
the example in this issue already. Elite is the worst first case: 43% ambiguous symbols, 16
overlays, and no licence.

One more thing worth noting: of the three byte-verified BeebAsm rebuilds out there (Moxon's Elite,
tom-seddon/exile_disassembly, kieranhj/thrust-disassembly), **not one passes `-d`/`-dd`/`-labels`**.
They all keep the `-v` listing instead. Adding the flag is one line per repo and would produce
symbol files for three of the most-wanted titles immediately, useful to b2, b-em and BeebEm as much
as to us. So: worth asking Mark Moxon about permission and about taking such a PR, and worth raising
a BeebAsm issue about exporting `name = value` constants, which would improve every consumer at
once.

None of the above touches the other half of this issue, the in-emulator RE tooling Rich described
(byte metadata, auto-labelling, recently-executed highlighting). That is still open and is probably
worth splitting out.

**Comment by mattgodbolt** [comment]
Parking this for now, with two decisions recorded so the research above does not have to be redone.

**Link, do not cannibalise.** Wherever it is reasonable, jsbeeb should point at the original site rather than copy its content in. That applies especially to Mark Moxon's disassemblies, which are an enormous body of careful work and are best read in their own context. Even where a small passage is quoted, the link to the original and its surrounding explanation is the more valuable thing to give the user, and the quote should not become a substitute for it.

That is a stronger version of what the research above concluded on technical grounds. Storing address to source-file and line, rather than copying comment text into a sidecar, happens to be both the format that can represent the real corpora and the one that keeps the reader pointed at the original. Those two arguments agreeing is a good sign.

**Not proceeding yet.** Simon M's beeb-vsc already does a good deal of this, including source-level debugging with jsbeeb embedded, so before building anything here the honest question is whether jsbeeb should do this at all or should defer to work that exists. That is worth answering before any code is written, not after.

So this stays open as a well-researched idea rather than a queued task. The state of things, for whoever picks it up:

- The symbol layer is uncontroversial and small: BeebAsm `-d`/`-dd` and VICE `al`, about twenty lines each, interoperable with b2, b-em and BeebEm.
- The annotated-source layer needs the source-map shape, not a flat sidecar, because neither large corpus fits one.
- Discovery by disc CRC32 needs no new machinery, since every load path already computes it.
- Exile is the sane first target if it ever starts. Elite is the worst one.
- The other half of this issue, the in-emulator reverse-engineering tooling, is untouched and probably wants splitting into its own issue.

**Comment by mattgodbolt** [comment]
Some discussions with @waitingforvsync about this wrt to baron and its output, maybe something something json blobs.
